decoder/decoder.go

```diff
@@ -87,6 +98,26 @@ func (s *Set) Decode(in []byte, label config.Label) ([]byte, error) {
 // DecodeLabels transforms eBPF map key bytes into a list of label values
 // according to configuration
 func (s *Set) DecodeLabels(in []byte, labels []config.Label) ([]string, error) {
+	cacheKey := maphash.Bytes(s.seed, in)
```

Comment:
I mentioned this internally, but it would be simpler to use `map[string]cacheValue{}` and `s.cache[string(in)]`. The Go compiler has a specific optimisation for this case.

https://github.com/golang/go/issues/3512

Reply:
current:

```sh
goos: linux
goarch: amd64
pkg: github.com/cloudflare/ebpf_exporter/v2/decoder
cpu: AMD Ryzen 9 3950X 16-Core Processor
BenchmarkCache/direct-32                 1097241              1155 ns/op
BenchmarkCache/cached-32                126773032                9.419 ns/op
```

with optimisation:

```sh
goos: linux
goarch: amd64
pkg: github.com/cloudflare/ebpf_exporter/v2/decoder
cpu: AMD Ryzen 9 3950X 16-Core Processor
BenchmarkCache/direct-32                 1102927              1032 ns/op
BenchmarkCache/cached-32                134451720                9.002 ns/op
```

```diff
diff --git a/decoder/decoder.go b/decoder/decoder.go
--- a/decoder/decoder.go
+++ b/decoder/decoder.go
@@ -1,7 +1,6 @@
 package decoder

 import (
-       "bytes"
        "errors"
        "fmt"
        "hash/maphash"
@@ -30,7 +29,7 @@
 type Set struct {
        mu       sync.Mutex
        decoders map[string]Decoder
-       cache    map[uint64]cacheValue
+       cache    map[string]cacheValue
        seed     maphash.Seed
 }

@@ -65,7 +64,7 @@
                        "syscall":      &Syscall{},
                        "uint":         &UInt{},
                },
-               cache: map[uint64]cacheValue{},
+               cache: map[string]cacheValue{},
                seed:  maphash.MakeSeed(),
        }, nil
 }
@@ -98,12 +97,8 @@
 // DecodeLabels transforms eBPF map key bytes into a list of label values
 // according to configuration
 func (s *Set) DecodeLabels(in []byte, labels []config.Label) ([]string, error) {
-       cacheKey := maphash.Bytes(s.seed, in)
-       if cached, ok := s.cache[cacheKey]; ok {
-               // Make sure that there is no collision here
-               if bytes.Equal(cached.in, in) {
-                       return cached.out, nil
-               }
+       if cached, ok := s.cache[string(in)]; ok {
+               return cached.out, nil
        }

        values, err := s.decodeLabels(in, labels)
@@ -111,7 +106,7 @@
                return nil, err
        }

-       s.cache[cacheKey] = cacheValue{in: in, out: values}
+       s.cache[string(in)] = cacheValue{in: in, out: values}

        return values, nil
 }
```